porch/controllers/packagevariantsets/pkg/controllers/packagevariantset/packagevariantset_controller_test.go

```diff
@@ -286,3 +289,77 @@ packageName:
 	},
 	}, result)
 }
+
+func TestGetSelectedObjects(t *testing.T) {
+	selectors := []api.Selector{{
+		APIVersion: "v1",
+		Kind:       "Pod",
+		Labels:     &metav1.LabelSelector{MatchLabels: map[string]string{"foo": "bar"}},
+	}}
+	reconciler := &PackageVariantSetReconciler{
+		Client:     new(fakeClient),
+		serializer: json.NewSerializerWithOptions(json.DefaultMetaFactory, nil, nil, json.SerializerOptions{Yaml: true}),
+	}
+	selectedObjects, err := reconciler.getSelectedObjects(context.Background(), selectors)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(selectedObjects))
+
+	expectedResId := resid.NewResIdWithNamespace(resid.NewGvk("", "v1", "Pod"), "my-pod-1", "")
+	obj, found := selectedObjects[expectedResId]
+	require.True(t, found)
+	require.Equal(t, `apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    abc: def
+    foo: bar
+  name: my-pod-1
+`, obj.MustString())
+}
+
+func TestObjectSet(t *testing.T) {
+	selectedObjects := map[resid.ResId]*kyaml.RNode{
+		resid.NewResIdWithNamespace(resid.NewGvk("", "v1", "Pod"), "my-pod-1", ""): kyaml.MustParse(`apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    repo: my-repo
+  name: downstream
+`),
+	}
+
+	target := &api.Target{
+		PackageName: &api.PackageName{
+			Name: &api.ValueOrFromField{FromField: "metadata.name"},
+		},
+		Objects: &api.ObjectSelector{
+			RepoName: &api.ValueOrFromField{FromField: "metadata.labels.repo"},
+		},
+	}
+
+	pvs := &PackageVariantSetReconciler{}
+	objectSet, err := pvs.objectSet(target, "upstream", selectedObjects)
+	require.NoError(t, err)
+	require.Equal(t, len(objectSet), 1)
+	require.Equal(t, pkgvarapi.Downstream{
+		Repo:    "my-repo",
+		Package: "downstream",
+	}, *objectSet[0])
+}
+
+func TestEnsurePackageVariants(t *testing.T) {
+	upstream := &pkgvarapi.Upstream{Repo: "up", Package: "up", Revision: "up"}
+	downstreams := []*pkgvarapi.Downstream{
+		{Repo: "dn-1", Package: "dn-1"},
+		{Repo: "dn-3", Package: "dn-3"},
+	}
+	pvs := &api.PackageVariantSet{ObjectMeta: metav1.ObjectMeta{Name: "my-pvs"}}
+
+	fc := &fakeClient{}
+	reconciler := &PackageVariantSetReconciler{Client: fc}
+	require.NoError(t, reconciler.ensurePackageVariants(context.Background(), upstream, downstreams, pvs))
+	require.Equal(t, []string{"listing objects",
```

Comment:
This way of testing the behavior seems a bit awkward. It is probably fine for now, but would be be clearer if we kept a an internal cache of objects and test the end state rather than behavior?

Reply:
done.